common/futures-semaphore/src/lib.rs

```diff
@@ -90,6 +94,28 @@ impl Semaphore {
         }
     }
 
+    /// Acquire an available permit from an owned semaphore.
+    ///
+    /// If there are no permits currently available, the future will wait until
+    /// a permit is released back to the semaphore.
+    pub async fn into_permit(self) -> Permit {
+        // Acquire a permit and then immediately "forget" it (drop it without
+        // returning the permit to the pool of available permits). We will
+        // manually add a new permit when our `Permit` RAII guard is dropped.
+        //
+        // We do this instead of holding tokio's `SemaphorePermit` in the RAII
+        // guard due to lifetime issues--tokio's `SemaphorePermit` needs a
+        // direct reference to the `Semaphore`; however, we want to send our
+        // `Permit` to other tasks or threads, so we need an `Arc` to the
+        // semaphore so the permit doesn't outlive the semaphore.
+        let permit = self.inner.semaphore.acquire().await;
+        permit.forget();
+
+        Permit {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+
```

Comment:
Lets simplify this code a bit by doing the following (but maybe improving the doc comment): 
```suggestion
    /// Consume `self` and call `acuire()`
    pub async fn into_permit(self) -> Permit {
        self.acquire().await
    }

```